### PR TITLE
Jim/FEQ-733/pull-the-changes-from-binary-websocket-api-automatically

### PR DESCRIPTION
**Pull request by jim-deriv**
## Changes
- Added a workflow to perform nightly syncs by `cloning`, and `committing` changes if any from `regentmarkets/binary-websocket-api` to `binary-com/deriv-api-docs/`
- Added a script to copy files from the `regentmarkets/binary-websocket-api` clone to the `binary-com/deriv-api-docs` clone

### CODE REVIEW

File: .github/workflows/sync.yml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+          git add -A ./_data
+          git add -A ./config/v3
+          git diff --name-only >> "$filename"
+
+      - name: Commit and push changes
+        run: |
+          filename="../logs/git_commit_$(date +'%Y-%m-%d-%H-%M-%S').txt"
+          cd deriv-api-docs/
+          if git commit -m "JSON Schema Update" 2> error.txt; then
+            git push origin master
```
Comment (balakrishna-deriv) [inline .github/workflows/sync.yml:60]: There's a branch protection rule. Not sure if we can push to master directly.

File: .github/scripts/get-web-socket-api-list.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ */
+const generateApiList = async () => {
+  console.log('Generating API list...');
+  const methods = [];
+  const yamlDocument = new YAML.Document();
+  try {
+    const filesAndFolders = await fs.promises.readdir(SOURCE_PATH, {
+      withFileTypes: true,
+    });
+    const methodNames = filesAndFolders
```
Comment (balakrishna-deriv) [inline .github/scripts/get-web-socket-api-list.js:23]:
```suggestion
    const apiNames = filesAndFolders
```

File: .github/workflows/sync.yml (hunk 1 above)
Comment (jim-deriv) [reply .github/workflows/sync.yml:60]: Alright @balakrishna-deriv, We'll verify this in my test repo

File: .github/scripts/get-web-socket-api-list.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      .filter((fileOrFolder) => fileOrFolder.isDirectory())
+      .map((directory) => directory.name);
+    for (const methodName of methodNames) {
+      const sendJsonPath = path.join(SOURCE_PATH, methodName, '/send.json');
+      const sendData = await fs.promises.readFile(sendJsonPath, 'utf8');
+      const send = JSON.parse(sendData);
+      if (!send.hidden) {
+        // Construct objects to be stored in the _data/v3.yml file
+        const title = send.title.replace(/ \(request\)$/, '');
+        methods.push({
```
Comment (balakrishna-deriv) [inline .github/scripts/get-web-socket-api-list.js:33]:
```suggestion
        apis.push({
```

File: .github/workflows/sync.yml (hunk 1 above)
Comment (jim-deriv) [reply .github/workflows/sync.yml]: Changed the strategy @balakrishna-deriv. Direct push was not possible

File: .github/workflows/sync.yml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+          commit-message: JSON Schema Update
+          title: JSON Schema Update
+          body: JSON Schema Update
+          branch: ${{ steps.generateBranchName.outputs.BRANCH }}
+          base: master
+          token: ${{ secrets.TEST_API_PR_PAT }} # TODO: Replace with correct Personal Access Token when provided
+          path: deriv-api-docs
+          delete-branch: true
+
+      - name: Merge Pull Request
```
Comment (balakrishna-deriv) [inline .github/workflows/sync.yml:63]: It looks like we have checks in place. I think we've to wait for those checks to be passed before merging.
image

File: .github/workflows/sync.yml [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+      - name: Create Pull Request
+        id: createPullRequest
+        uses: peter-evans/create-pull-request@v5
+        with:
+          commit-message: JSON Schema Update
+          title: JSON Schema Update
+          body: JSON Schema Update
+          branch: ${{ steps.generateBranchName.outputs.BRANCH }}
+          base: master
+          token: ${{ secrets.TEST_API_PR_PAT }} # TODO: Replace with correct Personal Access Token when provided
```
Comment (balakrishna-deriv) [inline .github/workflows/sync.yml:59]:
```suggestion
          token: ${{ secrets. GITHUB_TOKEN }} # TODO: Replace with correct Personal Access Token when provided
```
I think we can now create PRs from a github action. So, `GITHUB_TOKEN` should be sufficient. Please check with @ali-hosseini-deriv

Comment (jim-deriv) [reply .github/workflows/sync.yml:59]: Alright @balakrishna-deriv. The  `GITHUB_TOKEN ` didn't work in my test repo, but I'll use it. If it fails, then we can create one

**Review by balakrishna-deriv — APPROVED** [review]

File: .github/scripts/get-web-socket-api-list.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+const SOURCE_PATH = path.join(process.argv[2] || 'deriv-websockets-api', 'config/v3/');
+const ALL_CONFIG_PATH = path.join(BASE_PATH, '_data/');
+
+/*
+ *  This function copies all the api schema files from the regentmarkets/binary-websocket-api/config/v3 folder to the binary-com/deriv-api-docs/config/v3 folder,
+ *   whose hidden attribute is set to false.
+ *  It also generates the v3.yml file in the binary-com/deriv-api-docs/_data folder, that lists all the api calls
+ */
+const generateApiList = async () => {
+  console.log('Generating API list...');
```
Comment (niloofar-deriv) [inline .github/scripts/get-web-socket-api-list.js:16]: Do we need this console.log?

Comment (jim-deriv) [reply .github/scripts/get-web-socket-api-list.js:16]: @niloofar-deriv, This is just info that's printed when the action starts copying the files, I can remove it if you want me to.

Comment (niloofar-deriv) [reply .github/scripts/get-web-socket-api-list.js:16]: No that would be fine

**Review by niloofar-deriv — APPROVED** [review]

**Comment by jim-deriv** [comment]
Closing this PR, because the solution will be implemented from the `back-end` side

**Pull request closed without merging**